# Проект SQL по анализу базы данных сервиса для чтения книг по подписке.

## 
Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Наша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Наша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

### Импортируем библиотеки и пишем функцию для работы с SQL.

In [1]:
# импортируем библиотеки
import pandas as pd
import sqlalchemy as sa
# устанавливаем параметры
db_config = {
'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db' # название базы данных
}
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = sa.create_engine(connection_string, connect_args={'sslmode':'require'})
# чтобы выполнить SQL-запрос, пишем функцию с использованием Pandas
def get_sql_data(query:str, engine:sa.engine.base.Engine=engine) -> pd.DataFrame:
    '''Открываем соединение, получаем данные из sql, закрываем соединение'''
    with engine.connect() as con:
        return pd.read_sql(sql=sa.text(query), con = con)


### Знакомимся с данными.

#### Описание данных

Таблица books

Содержит данные о книгах:
* book_id — идентификатор книги;
* author_id — идентификатор автора;
* title — название книги;
* num_pages — количество страниц;
* publication_date — дата публикации книги;
* publisher_id — идентификатор издателя.

Таблица authors

Содержит данные об авторах:
* author_id — идентификатор автора;
* author — имя автора.

Таблица publishers

Содержит данные об издательствах:
* publisher_id — идентификатор издательства;
* publisher — название издательства;

Таблица ratings

Содержит данные о пользовательских оценках книг:
* rating_id — идентификатор оценки;
* book_id — идентификатор книги;
* username — имя пользователя, оставившего оценку;
* rating — оценка книги.

Таблица reviews

Содержит данные о пользовательских обзорах:
* review_id — идентификатор обзора;
* book_id — идентификатор книги;
* username — имя автора обзора;
* text — текст обзора.

#### Структура данных.

In [2]:
from IPython.display import Image
from IPython.core.display import HTML 
Image(url= "https://pictures.s3.yandex.net/resources/scheme_1589269096.png")

#### Таблица books.

Выведем первые 5 строк из таблицы books и посчитаем количество строк.

In [3]:
# формируем запрос и выводим данные
query = '''
SELECT * 
FROM books 
LIMIT 5'''
get_sql_data(query)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [4]:
query = '''
SELECT COUNT(*)
FROM books 
LIMIT 5
'''
get_sql_data(query)

count
0   1000

В таблице books 1000 строк.

#### Таблица authors.

Выведем первые 5 строк из таблицы authors и посчитаем количество строк.

In [5]:
query = '''
SELECT * 
FROM authors 
LIMIT 5'''
get_sql_data(query)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [6]:
query = '''
SELECT COUNT(*)
FROM authors 
LIMIT 5
'''
get_sql_data(query)

count
0    636

В таблице authors 636 строк.

#### Таблица publishers.

Выведем первые 5 строк из таблицы publishers и посчитаем количество строк.

In [7]:
query = '''
SELECT * 
FROM publishers 
LIMIT 5'''
get_sql_data(query)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [8]:
query = '''
SELECT COUNT(*)
FROM publishers 
LIMIT 5
'''
get_sql_data(query)

count
0    340

В таблице publishers 340 строк.

#### Таблица ratings.

Выведем первые 5 строк из таблицы ratings и посчитаем количество строк.

In [9]:
query = '''
SELECT * 
FROM ratings 
LIMIT 5'''
get_sql_data(query)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [10]:
query = '''
SELECT COUNT(*)
FROM ratings 
LIMIT 5
'''
get_sql_data(query)

count
0   6456

В таблице ratings 6456 строк.

#### Таблица reviews.

Выведем первые 5 строк из таблицы reviews и посчитаем количество строк.

In [11]:
query = '''
SELECT * 
FROM reviews 
LIMIT 5'''
get_sql_data(query)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [12]:
query = '''
SELECT COUNT(*)
FROM reviews 
LIMIT 5
'''
get_sql_data(query)

count
0   2793

В таблице reviews 2793 строки.

### Исследуем данные.

#### Посчитаем, сколько книг вышло после 1 января 2000 года.

In [13]:
query = '''
SELECT COUNT(*)
FROM books 
WHERE publication_date::date > '2000-01-01'
'''
get_sql_data(query)

count
0    819

После 1 января 2000 года вышло 819 книг.

#### Для каждой книги посчитаем количество обзоров и среднюю оценку.

In [14]:
query = '''
SELECT b.title,
       ROUND(AVG(ra.rating),2) AS avg_rating,
       COUNT(r.review_id) AS count_rev
FROM books AS b
LEFT JOIN reviews AS r ON b.book_id=r.book_id
LEFT JOIN ratings AS ra ON b.book_id=ra.book_id
GROUP BY title
ORDER BY count_rev DESC, avg_rating DESC
LIMIT 10;
'''
get_sql_data(query)

title  avg_rating  count_rev
0                            Twilight (Twilight  #1)        3.66       1120
1                The Hobbit  or There and Back Again        4.13        528
2                             The Catcher in the Rye        3.83        516
3  Harry Potter and the Prisoner of Azkaban (Harr...        4.41        492
4  Harry Potter and the Chamber of Secrets (Harry...        4.29        480
5               Angels & Demons (Robert Langdon  #1)        3.68        420
6  Harry Potter and the Order of the Phoenix (Har...        4.19        375
7  The Lightning Thief (Percy Jackson and the Oly...        4.08        372
8  The Fellowship of the Ring (The Lord of the Ri...        4.39        370
9                                        Animal Farm        3.73        370

Самая обозреваемая книга - Сумерки, часть 1.

На втором месте Хоббит, или Туда и обратно.

Третье место - Над пропастью во ржи.

В целом можно сказать что самые обозреваемые книги в основном из жанра фэнтези - 7 из 10.

#### Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры.

In [15]:
query = '''
SELECT p.publisher,
       b.publisher_id,
       COUNT(b.book_id)
FROM books AS b
LEFT JOIN publishers AS p ON b.publisher_id=p.publisher_id
WHERE num_pages > 50
GROUP BY p.publisher,b.publisher_id
ORDER BY COUNT(b.book_id) DESC
LIMIT 5;
'''
get_sql_data(query)

publisher  publisher_id  count
0             Penguin Books           212     42
1                   Vintage           309     31
2  Grand Central Publishing           116     25
3          Penguin Classics           217     24
4                    Bantam            35     19

Издательство, которое выпустило наибольшее число книг толще 50 страниц - Penguin Books, 42 книги.

#### Определим автора с самой высокой средней оценкой книг — учитываем только книги с 50 и более оценками.

In [16]:
query = '''
SELECT a.author,
       ROUND(AVG(ra.rating),2)
FROM books AS b
LEFT JOIN authors AS a ON b.author_id=a.author_id
LEFT JOIN ratings AS ra ON b.book_id=ra.book_id
GROUP BY b.book_id,a.author,b.author_id
HAVING COUNT(rating_id) >= 50
ORDER BY AVG(ra.rating) DESC
LIMIT 1;
'''
get_sql_data(query)

author  round
0  J.K. Rowling/Mary GrandPré   4.41

Автор с самой высокой средней оценкой книг - J.K. Rowling/Mary GrandPré, среди книг с 50 и более оценками.

#### Посчитаем среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [17]:
query = '''
SELECT AVG(slice.rev_count)
FROM (SELECT username,
             COUNT(review_id) AS rev_count
      FROM reviews
      WHERE username IN (SELECT username
                         FROM ratings
                         GROUP BY username
                         HAVING COUNT(rating_id) > 48)
     GROUP BY username) AS slice
'''
get_sql_data(query)

avg
0  24.0

Cреднее количество обзоров от пользователей, которые поставили больше 48 оценок - 24.